backend/description.go

```diff
@@ -0,0 +1,177 @@
+//
+// We will never require a CLA to submit a patch. All contributions follow the
+// `inbound == outbound` rule.
+//
+// This is not an official Amazon product. Amazon does not offer support for
+// this project.
+
+// TODO: should this be a subpackage?
+package backend
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"sort"
+	"strings"
+
+	"github.com/awslabs/yesiscan/interfaces"
+	"github.com/awslabs/yesiscan/util/errwrap"
+	"github.com/awslabs/yesiscan/util/licenses"
+)
+
+const (
+	// DescriptionMaxBytesLine sets a larger maximum for file line scanning than
+	// the default of bufio.MaxScanTokenSize which is sort of small.
+	DescriptionMaxBytesLine = 1024 * 1024 * 8 // 8 MiB
+
+	// DescriptionLicensePrefix is the string we look for when trying to find a
+	// license.
+	DescriptionLicensePrefix = "License: "
+
+	// DescriptionFilename is the file name used by the R metadata files.
+	DescriptionFilename = "DESCRIPTION"
+)
+
+// Description is a backend for DESCRIPTION files which store R package
+// metadata. We are getting the license names from the License field in the text
+// file.
+type Description struct {
+	Debug bool
+	Logf  func(format string, v ...interface{})
+}
+
+// String method returns the name of the backend.
+func (obj *Description) String() string {
+	return "description"
+}
+
+// ScanData method is used to extract license ids from data and return licenses
+// based on the license ids.
+func (obj *Description) ScanData(ctx context.Context, data []byte, info *interfaces.Info) (*interfaces.Result, error) {
+	// This check is taking place with the assumption that the file that will be
+	// scanned will have to be named "DESCRIPTION".
+	if info.FileInfo.Name() != DescriptionFilename {
+		return nil, nil // skip
+	}
+	if info.FileInfo.IsDir() {
+		return nil, nil // skip
+	}
+	if len(data) == 0 {
+		return nil, nil // skip
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	licenseMap := make(map[string]struct{})
+	reader := bytes.NewReader(data)
+	scanner := bufio.NewScanner(reader)
+	buf := []byte{}                              // create a buffer for very long lines
+	scanner.Buffer(buf, DescriptionMaxBytesLine) // set the max size of that buffer
+	for scanner.Scan() {
+		// In an effort to short-circuit things if needed, we run a check ourselves
+		// and break out early if we see that we have cancelled early.
+		select {
+		case <-ctx.Done():
+			return nil, errwrap.Wrapf(ctx.Err(), "scanner ended early")
+		default:
+		}
+
+		s := scanner.Text() // newlines will be stripped here
+		if !strings.HasPrefix(s, DescriptionLicensePrefix) {
+			continue
+		}
+		license := s[len(DescriptionLicensePrefix):len(s)]
+		if license == "" {
+			// TODO: should we warn here?
+			continue
+		}
+		// XXX: These are the possibilities of useless strings that need to be removed
+		// to get licenses so far.
+		// source: https://cran.rstudio.com/doc/manuals/r-devel/R-exts.html#Licensing
+		replacer := strings.NewReplacer(
```

Comment:
Good try, but this sub-parser is too trivial for what we need. Can you propose (design in words) a better alternative?

We talked about this during our 1:1 -- the plan is to write tests and to split this out into a separate function and we'll look back at it after the tests are done! \o/